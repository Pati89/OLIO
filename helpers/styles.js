// here we can add theme, fontSizes or spacing sizes to use same all over the page

export const breakpoints = {
  sm: 640,
  md: 896,
  lg: 1092,
};

export const media = {
  /* small size and up */
  sm: `@media only screen and (min-width: ${breakpoints.xs}px)`,
  /* medium size and up */
  md: `@media only screen and (min-width: ${breakpoints.sm}px)`,
  /* large size and up */
  lg: `@media (min-width: ${breakpoints.md}px)`,
};
