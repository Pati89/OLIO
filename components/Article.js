import PropTypes from "prop-types";
import Image from "next/image";
import styled from "@emotion/styled";
import { media } from "../helpers/styles";

const S_a_card = styled.a`
  width: 100%;
  height: 100%;
  display: flex;
  flex-direction: row;
  text-align: left;
  color: inherit;
  text-decoration: none;
  border: 1px solid black;
  border-radius: 10px;
  overflow: hidden;
  cursor: pointer;

  ${media.md} {
    flex-direction: column;
  }
`;

const S_div_image = styled.div`
  position: relative;
  width: 35%;
  height: 100%;
  max-height: 300px;

  ${media.md} {
    width: 100%;
    height: 300px;
    max-height: 300px;
  }
`;

const S_div_card_content = styled.div`
  width: 65%;
  margin: 20px;

  ${media.md} {
    width: auto;
  }
`;

const S_p_title = styled.p`
  color: inherit;
  font-weight: bold;
  transition: color 0.15s ease;
  margin-top: 0;

  &:hover,
  &:focus,
  &:active {
    color: #0070f3;
  }
`;

const S_div_user_container = styled.div`
  display: flex;
  justify-content: space-between;
  align-items: center;
`;

const S_div_avatar = styled.div`
  display: flex;
  position: relative;
  width: 30px;
  height: 30px;
  border-radius: 15px;
  overflow: hidden;
`;

const S_div_user = styled.div`
  display: flex;
  align-items: center;
`;

const S_div_status = styled.div`
  display: flex;
  align-items: center;
  background-color: ${({ status }) => (status === "active" ? "green" : "grey")};
  color: white;
  padding: 5px;
  border-radius: 10px;
  font-size: 0.75rem;
`;

const S_span_user = styled.span`
  margin-left: ${({ avatarSrc }) => (avatarSrc ? "10px" : 0)};
`;

const Article = (props) => {
  const { imageSrc, imageAlt, title, avatarSrc, avatarAlt, user, status } =
    props;

  return (
    <S_a_card onClick={() => {}}>
      <S_div_image>
        {/* I'm using here next.js image. This is already lazy loaded */}
        <Image className="image" src={imageSrc} alt={imageAlt} layout="fill" />
      </S_div_image>
      <S_div_card_content>
        <S_p_title>{title}</S_p_title>
        <S_div_user_container>
          <S_div_user>
            {avatarSrc && (
              <S_div_avatar>
                <Image src={avatarSrc} alt={avatarAlt} layout="fill" />
              </S_div_avatar>
            )}
            <S_span_user avatarSrc={avatarSrc}>{user}</S_span_user>
          </S_div_user>
          <S_div_status status={status}>{status}</S_div_status>
        </S_div_user_container>
      </S_div_card_content>
    </S_a_card>
  );
};

Article.propTypes = {
  imageSrc: PropTypes.string,
  imageAlt: PropTypes.string,
  title: PropTypes.string,
  avatarSrc: PropTypes.string,
  avatarAlt: PropTypes.string,
  user: PropTypes.string,
  status: PropTypes.string,
};

export default Article;
