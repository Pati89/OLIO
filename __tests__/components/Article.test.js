import Article from "../../components/Article";
import { render } from "../../helpers/test-utils";

const props = {
  imageSrc:
    "https://cdn.olioex.com/uploads/photo/file/Y9d--jZv7f0GoLDn3SbH6A/medium_image.jpg",
  title: "test title",
  user: "Olio user",
};

describe("Article", () => {
  it("should render a Article", () => {
    const { getByTestId } = render(<Article {...props} />);
    expect(getByTestId("article")).toBeTruthy();
  });

  it("should not render a Article", () => {
    const { queryByTestId } = render(<Article />);
    expect(queryByTestId("article")).toBeFalsy();
  });

  it("should render a title", () => {
    const { getByText } = render(<Article {...props} />);
    expect(getByText("test title")).toBeTruthy();
  });

  it("should render user", () => {
    const { getByText } = render(<Article {...props} />);
    expect(getByText("Olio user")).toBeTruthy();
  });

  it("should not render a avatar", () => {
    const { queryByTestId } = render(<Article {...props} />);
    expect(queryByTestId("avatar")).toBeFalsy();
  });

  it("should render a avatar", () => {
    const { queryByTestId } = render(
      <Article
        {...props}
        avatarSrc="https://cdn.olioex.com/uploads/avatar/file/oZq8DF3dzLEi3Fnf4XxMrg/small_image.jpg"
      />
    );
    expect(queryByTestId("avatar")).toBeTruthy();
  });

  it("should render active status", () => {
    const { getByText, getByTestId } = render(
      <Article {...props} status="active" />
    );
    expect(getByText("active")).toBeTruthy();
    expect(getByTestId("status")).toHaveStyleRule("background-color", "green");
  });

  it("should render other status", () => {
    const { getByText, getByTestId } = render(
      <Article {...props} status="inactive" />
    );
    expect(getByText("inactive")).toBeTruthy();
    expect(getByTestId("status")).toHaveStyleRule("background-color", "grey");
  });

  it("should show viewed badge", () => {
    const { getByText, getByTestId } = render(
      <Article {...props} status="inactive" />
    );
    expect(getByText("inactive")).toBeTruthy();
    expect(getByTestId("status")).toHaveStyleRule("background-color", "grey");
  });

  it("should render a Article match snapshot", () => {
    const { container } = render(<Article {...props} />);
    expect(container).toMatchSnapshot();
  });
});
