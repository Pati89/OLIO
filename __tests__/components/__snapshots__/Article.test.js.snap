// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Article should render a Article match snapshot 1`] = `
.emotion-0 {
  width: 100%;
  height: 100%;
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: row;
  -ms-flex-direction: row;
  flex-direction: row;
  text-align: left;
  color: inherit;
  -webkit-text-decoration: none;
  text-decoration: none;
  border-radius: 10px;
  overflow: hidden;
  cursor: pointer;
  background-color: white;
}

@media only screen and (min-width: 640px) {
  .emotion-0 {
    -webkit-flex-direction: column;
    -ms-flex-direction: column;
    flex-direction: column;
  }
}

.emotion-1 {
  background-color: lightgray;
  position: relative;
  width: 35%;
  min-height: 120px;
  max-height: 300px;
  opacity: 1;
}

@media only screen and (min-width: 640px) {
  .emotion-1 {
    width: 100%;
    height: 300px;
    max-height: 300px;
  }
}

.emotion-2 {
  width: 65%;
  margin: 20px;
  opacity: 1;
}

@media only screen and (min-width: 640px) {
  .emotion-2 {
    width: auto;
  }
}

.emotion-3 {
  color: inherit;
  font-weight: bold;
  -webkit-transition: color 0.15s ease;
  transition: color 0.15s ease;
  margin-top: 0;
}

.emotion-3:hover,
.emotion-3:focus,
.emotion-3:active {
  color: #0070f3;
}

.emotion-4 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-box-pack: justify;
  -webkit-justify-content: space-between;
  justify-content: space-between;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-5 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
}

.emotion-6 {
  margin-left: 0;
}

.emotion-7 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-align-items: center;
  -webkit-box-align: center;
  -ms-flex-align: center;
  align-items: center;
  background-color: grey;
  color: white;
  padding: 5px;
  border-radius: 10px;
  font-size: 0.75rem;
}

<div>
  <a
    class="emotion-0"
    href="/article/undefined"
  >
    <div
      class="emotion-1"
      data-testid="article"
    >
      <span
        style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
      >
        <img
          alt="test title"
          class="image"
          data-nimg="fill"
          decoding="async"
          src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
          style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%;"
        />
        <noscript />
      </span>
    </div>
    <div
      class="emotion-2"
    >
      <p
        class="emotion-3"
      >
        test title
      </p>
      <div
        class="emotion-4"
      >
        <div
          class="emotion-5"
        >
          <span
            class="emotion-6"
          >
            Olio user
          </span>
        </div>
        <div
          class="emotion-7"
          data-testid="status"
        />
      </div>
    </div>
  </a>
</div>
`;
